### ISSUE TITLE
I want support for static hosts file (/etc/hosts) to override DNS rather than block hosts

### ISSUE THREAD

**Issue by dzrdmwjx** [blokadaorg/blokada#563]
All popular operating systems (including mobile) support a hosts config file (etc/hosts) to define static host entries for the DNS resolver. If the queried host is found to have an entry in this file, no DNS lookup is done and the IP address specified in this file is returned instead.

Although this mechanism was often used for blocking ads by specifying an invalid IP address for the ad-servers, the primary function of the hosts config is to override DNS and not to block hosts. Now blokada has a much better way of blocking ad servers using block-lists because it controls the DNS lookups, but it doesn't provide a way to specify static host entries with IP addresses for the name resolution (that the earlier solution using /etc/hosts provided).

Since blokada effectively takes over DNS resolution when it's active, it would be great if support for static hosts entries is also added to its functionality. This would be very useful when the user wants to redirect certain domains to an alternate IP address rather than block them.

**Comment by peterroth** [comment]
Hi,

Blokada, to save battery and memory, doesn't check the IP address in the host list. In earlier versions Blokada redirected the IP, but for a long time now, you don't have to provide an IP, simply sends a "hostname not found" message to the app if a hostname is on a blacklist.
If this feature is implemented, Blokada has to go through the lists and check if that's redirected to other than 0.0.0.0 address, causing slower blocking and more memory consumption.

**Comment by dzrdmwjx** [comment]
Not necessarily... This feature can be implemented in a cleaner way that doesn't have this problem.

I'm proposing a distinction be made between a block-list and static-hosts configuration.
None of the existing block-list related functionality needs to change because the way it's being done is certainly better.

In addition, a new configuration option should be added to specify static host entries. This could be a standard format "hosts" file or a list of hosts/IPs entered through UI. These hosts should not be blocked, but resolved based on the static configuration rather than downstream DNS lookup.

So the application flow would be something like this:
- if the queried host is in block-list, send "domain-not-found"
- else if the queried host in in static config, send corresponding IP address
- else send the query to the downstream DNS server to resolve

So the only performance penalty is the additional check if a static configuration exists for the queried host. This would be insignificant if this configuration is read from an in-memory map (practically, there would be only a handful of these entries, as opposed to thousands for the block-list).

**Comment by immanuelfodor** [comment]
This feature would be fantastic for local/LAN web development/testing on non-rooted phones! 

I'd switch the order in this condition though: 

- if the queried host in in static config, send corresponding IP address
- else if the queried host is in block-list, send "domain-not-found"
- ... 

I suppose the static config would be empty most of the time or only has a few entries. This way the performance would be better to check that first rather than several hundred thousand blocked entries.

**Comment by alexa-v2** [comment]
Would caching the static hosts file mitigate performance concerns much? As long as it's not too large ofc

**Comment by jonas2k** [comment]
I'd love to see a static host list too. At the moment, I have to disable Blokada every time I want to access resources like web applications hosted in my local network, since their DNS names can't be resolved otherwise. Calling IP addresses directly works, but isn't an option for me personally. Or is there another possibility to achieve this?

Keep up the good work, Blokada is awesome! :)

**Comment by vdbhb59** [comment]
While on this topic, may I know, where does blokada currently store the hosts file upon selecting and syncing/updating lists? Does it override or update the /etc/hosts? A path would be great to know.

**Comment by peterroth** [comment]
As Blokada is a non-root solution, it can't, therefore it doesn't override the /system/etc/hosts. It keeps the active filters (hostnames) in memory and the selectes lists footprint in the paperdb.io file.
